### PR TITLE
[ci] [python-package] move tool configuration to pyproject.toml

### PR DESCRIPTION
**Pull request by jameslamb**
Contributes to #3867.
Contributes to #4308.
Contributes to #5061.

Proposes moving configuration for all Python linting tools that support it into `pyproject.toml` (and for `flake8`, which doesn't support it per https://github.com/PyCQA/flake8/issues/234, into `setup.cfg`).

Also adds configuration to skip running linters over the `lightgbm-python/` directory (new local build directory introduced in #5837), to help with local development.

### Notes for Reviewers

Many Python tools have been moving in this direction, ever since PEP 518 was accepted back in 2016.

From the PEP ([link](https://peps.python.org/pep-0518/#tool-table)) ...

> *The `[tool]` table is where any tool related to your Python project, not just build tools, can have users specify configuration data as long as they use a sub-table within `[tool]`, e.g. the `flit` tool would store its configuration in `[tool.flit]`.*

And projects *using those tools* have followed. For example, https://github.com/dmlc/xgboost/pull/9112.

### Benefits of this Change

Sets up `pyproject.toml` and `setup.cfg` as sources of configuration for the Python project, reducing the size of the diff to be reviewed in #5759.

Puts tool configuration into a format used by many other Python projects, making it a bit easier (I hope) for new contributors to work on LightGBM.

### CODE REVIEW

**Review by UriyaHarpeness — APPROVED** [review]

File: .ci/lint-python.sh [hunk 1]
```diff
@@ -1,41 +1,32 @@
 #!/bin/sh
 
-DIR_TO_CHECK=${1}
-
 echo "running flake8"
 flake8 \
+    --config=./python-package/setup.cfg \
     --ignore=E501,W503 \
     --exclude=./.nuget,./external_libs,./python-package/build,./python-package/compile \
```
Comment (jmoralez) [inline .ci/lint-python.sh:7]: Doesn't setting the config mean we can remove these two lines?

Comment (jameslamb) [reply .ci/lint-python.sh:7]: ah yes! Missed this one, thank you. I'll push a commit in a bit removing them.

Comment (jameslamb) [reply .ci/lint-python.sh]: fixed in https://github.com/microsoft/LightGBM/pull/5872/commits/03e624bea6c4bfba8b531b40d124ec0ffd56eb67, thanks for catching that!

**Review by jmoralez — APPROVED** [review]

**Comment by jameslamb** [comment]
closing and re-opening to try to get rid of that `license/cla` check 🙃 

Screen Shot 2023-05-08 at 3 59 08 PM

**Pull request merged**